src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs

```diff
@@ -781,8 +787,16 @@ public bool StartAsyncOperation(SocketAsyncContext context, TOperation operation
                                 }
 
                                 _tail = operation;
-
                                 Trace(context, $"Leave, enqueued {IdOf(operation)}");
+
+                                // Now that the object is enqueued, hook up cancellation.
+                                // Note that it's possible the call to register itself could
+                                // call TryCancel, so we do this after the op is fully enqueued.
+                                if (cancellationToken.CanBeCanceled)
+                                {
+                                    operation.CancellationRegistration = cancellationToken.UnsafeRegister(s => ((TOperation)s).TryCancel(), operation);
```

Comment:
@stephentoub a canceled operation will remain in the queue until it is removed at the next event or the queue gets stopped.

Reply:
Yes. Is that a problem?

Reply:
Not functionally. It will be alive for longer and maybe keeping some other things alive.

Reply:
> It will be alive for longer and maybe keeping some other things alive.

Yes, but then again so are the Socket and the SocketAsyncEventArgs. In comparison this shouldn't keep alive much. And if you're canceling the operation, there's a really high likelihood you're also either tearing everything down or about to do something else that will revisit the queue.